File: 2023/day22/java/Solution.java
import java.io.*;
import java.nio.file.*;
import java.util.*;

/**
 * Day 22: Sand Slabs - 3D falling bricks simulation.
 */
public class Solution {

    public static void main(String[] args) throws IOException {
        Path inputPath = Path.of(args.length > 0 ? args[0] : "../input.txt");
        List<int[]> bricks = parseInput(inputPath);

        System.out.println("Part 1: " + part1(bricks));
        System.out.println("Part 2: " + part2(bricks));
    }

    static List<int[]> parseInput(Path path) throws IOException {
        List<int[]> bricks = new ArrayList<>();
        for (String line : Files.readAllLines(path)) {
            String[] parts = line.split("~");
            String[] left = parts[0].split(",");
            String[] right = parts[1].split(",");

            int x1 = Integer.parseInt(left[0]);
            int y1 = Integer.parseInt(left[1]);
            int z1 = Integer.parseInt(left[2]);
            int x2 = Integer.parseInt(right[0]);
            int y2 = Integer.parseInt(right[1]);
            int z2 = Integer.parseInt(right[2]);

            // Ensure z1 <= z2 for consistent processing
            if (z1 > z2) {
                int tx = x1, ty = y1, tz = z1;
                x1 = x2; y1 = y2; z1 = z2;
                x2 = tx; y2 = ty; z2 = tz;
            }

            bricks.add(new int[]{x1, y1, z1, x2, y2, z2});
        }
        return bricks;
    }

    static class SettleResult {
        int[][] settled;
        Map<Integer, Set<Integer>> supports;      // brick i -> bricks it supports (above)
        Map<Integer, Set<Integer>> supporters;    // brick i -> bricks that support it (below)

        SettleResult(int n) {
            settled = new int[n][];
            supports = new HashMap<>();
            supporters = new HashMap<>();
            for (int i = 0; i < n; i++) {
                supports.put(i, new HashSet<>());
                supporters.put(i, new HashSet<>());
            }
        }
    }

    static SettleResult settleBricks(List<int[]> bricks) {
        int n = bricks.size();
        SettleResult result = new SettleResult(n);

        // Create indexed list and sort by minimum z coordinate
        Integer[] order = new Integer[n];
        for (int i = 0; i < n; i++) order[i] = i;
        Arrays.sort(order, (a, b) -> Integer.compare(bricks.get(a)[2], bricks.get(b)[2]));

        // Track occupied cells: (x, y, z) -> brick index
        Map<Long, Integer> occupied = new HashMap<>();

        for (int origIdx : order) {
            int[] brick = bricks.get(origIdx);
            int x1 = brick[0], y1 = brick[1], z1 = brick[2];
            int x2 = brick[3], y2 = brick[4], z2 = brick[5];

            // Find the lowest z where this brick can rest
            int drop = z1 - 1;  // Maximum drop (to z=1)

            // Get xy footprint of this brick
            for (int x = Math.min(x1, x2); x <= Math.max(x1, x2); x++) {
                for (int y = Math.min(y1, y2); y <= Math.max(y1, y2); y++) {
                    // Check each z level below the brick
                    for (int z = z1 - 1; z > 0; z--) {
                        if (occupied.containsKey(key(x, y, z))) {
                            drop = Math.min(drop, z1 - z - 1);
                            break;
                        }
                    }
                }
            }

            // Drop the brick
            int newZ1 = z1 - drop;
            int newZ2 = z2 - drop;
            result.settled[origIdx] = new int[]{x1, y1, newZ1, x2, y2, newZ2};

            // Mark cells as occupied and find supporters
            for (int x = Math.min(x1, x2); x <= Math.max(x1, x2); x++) {
                for (int y = Math.min(y1, y2); y <= Math.max(y1, y2); y++) {
                    // Check if there's a brick directly below
                    long belowKey = key(x, y, newZ1 - 1);
                    if (occupied.containsKey(belowKey)) {
                        int supporterIdx = occupied.get(belowKey);
                        result.supporters.get(origIdx).add(supporterIdx);
                        result.supports.get(supporterIdx).add(origIdx);
                    }

                    // Mark all cells of this brick as occupied
                    for (int z = newZ1; z <= newZ2; z++) {
                        occupied.put(key(x, y, z), origIdx);
                    }
                }
            }
        }

        return result;
    }

    static long key(int x, int y, int z) {
        // Pack x, y, z into a single long (assuming coords fit in ~20 bits each)
        return ((long)x << 40) | ((long)y << 20) | z;
    }

    static int part1(List<int[]> bricks) {
        SettleResult sr = settleBricks(bricks);

        int safeCount = 0;
        for (int i = 0; i < bricks.size(); i++) {
            // Brick i can be safely removed if every brick it supports
            // has at least one other supporter
            boolean canRemove = true;
            for (int supported : sr.supports.get(i)) {
                if (sr.supporters.get(supported).size() == 1) {
                    canRemove = false;
                    break;
                }
            }
            if (canRemove) {
                safeCount++;
            }
        }

        return safeCount;
    }

    static int part2(List<int[]> bricks) {
        SettleResult sr = settleBricks(bricks);

        int totalFalls = 0;

        for (int i = 0; i < bricks.size(); i++) {
            // Simulate removing brick i and count chain reaction
            // BFS to find all bricks that would fall
            Set<Integer> falling = new HashSet<>();
            falling.add(i);
            Deque<Integer> queue = new ArrayDeque<>();
            queue.add(i);

            while (!queue.isEmpty()) {
                int brick = queue.poll();

                // Check all bricks that this brick supports
                for (int supported : sr.supports.get(brick)) {
                    if (falling.contains(supported)) {
                        continue;
                    }

                    // This brick falls if all its supporters have fallen
                    if (falling.containsAll(sr.supporters.get(supported))) {
                        falling.add(supported);
                        queue.add(supported);
                    }
                }
            }

            // Don't count the initial brick we removed
            totalFalls += falling.size() - 1;
        }

        return totalFalls;
    }
}

File: 2023/day22/cpp/solution.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <map>
#include <set>
#include <queue>
#include <algorithm>
#include <tuple>

struct Brick {
    int x1, y1, z1, x2, y2, z2;
};

std::vector<Brick> parse_input(const std::string& filename) {
    std::vector<Brick> bricks;
    std::ifstream file(filename);
    std::string line;

    while (std::getline(file, line)) {
        if (line.empty()) continue;

        // Parse "x1,y1,z1~x2,y2,z2"
        size_t tilde = line.find('~');
        std::string left = line.substr(0, tilde);
        std::string right = line.substr(tilde + 1);

        Brick b;
        char comma;
        std::istringstream lss(left);
        lss >> b.x1 >> comma >> b.y1 >> comma >> b.z1;
        std::istringstream rss(right);
        rss >> b.x2 >> comma >> b.y2 >> comma >> b.z2;

        // Ensure z1 <= z2 for consistent processing
        if (b.z1 > b.z2) {
            std::swap(b.x1, b.x2);
            std::swap(b.y1, b.y2);
            std::swap(b.z1, b.z2);
        }

        bricks.push_back(b);
    }

    return bricks;
}

struct SettleResult {
    std::vector<Brick> settled;
    std::vector<std::set<int>> supports;    // supports[i] = bricks that i supports (above)
    std::vector<std::set<int>> supporters;  // supporters[i] = bricks that support i (below)
};

SettleResult settle_bricks(std::vector<Brick>& bricks) {
    int n = bricks.size();

    // Create indices and sort by minimum z
    std::vector<int> order(n);
    for (int i = 0; i < n; i++) order[i] = i;
    std::sort(order.begin(), order.end(), [&](int a, int b) {
        return std::min(bricks[a].z1, bricks[a].z2) < std::min(bricks[b].z1, bricks[b].z2);
    });

    // Track occupied cells: (x, y, z) -> brick index
    std::map<std::tuple<int, int, int>, int> occupied;

    std::vector<Brick> settled(n);
    std::vector<std::set<int>> supports(n);
    std::vector<std::set<int>> supporters(n);

    for (int orig_idx : order) {
        Brick& brick = bricks[orig_idx];
        int x1 = brick.x1, y1 = brick.y1, z1 = brick.z1;
        int x2 = brick.x2, y2 = brick.y2, z2 = brick.z2;

        // Find the lowest z where this brick can rest
        int drop = z1 - 1;  // Maximum drop (to z=1)

        // Get xy footprint and find max obstacle
        for (int x = std::min(x1, x2); x <= std::max(x1, x2); x++) {
            for (int y = std::min(y1, y2); y <= std::max(y1, y2); y++) {
                // Check each z level below the brick
                for (int z = z1 - 1; z >= 1; z--) {
                    auto key = std::make_tuple(x, y, z);
                    if (occupied.find(key) != occupied.end()) {
                        drop = std::min(drop, z1 - z - 1);
                        break;
                    }
                }
            }
        }

        // Drop the brick
        int new_z1 = z1 - drop;
        int new_z2 = z2 - drop;
        Brick new_brick = {x1, y1, new_z1, x2, y2, new_z2};
        settled[orig_idx] = new_brick;

        // Mark cells as occupied and find supporters
        for (int x = std::min(x1, x2); x <= std::max(x1, x2); x++) {
            for (int y = std::min(y1, y2); y <= std::max(y1, y2); y++) {
                // Check if there's a brick directly below
                auto below_key = std::make_tuple(x, y, new_z1 - 1);
                auto it = occupied.find(below_key);
                if (it != occupied.end()) {
                    int supporter_idx = it->second;
                    supporters[orig_idx].insert(supporter_idx);
                    supports[supporter_idx].insert(orig_idx);
                }

                // Mark all cells of this brick as occupied
                for (int z = new_z1; z <= new_z2; z++) {
                    occupied[std::make_tuple(x, y, z)] = orig_idx;
                }
            }
        }
    }

    return {settled, supports, supporters};
}

int part1(std::vector<Brick>& bricks) {
    auto result = settle_bricks(bricks);
    auto& supports = result.supports;
    auto& supporters = result.supporters;

    int safe_count = 0;
    int n = bricks.size();

    for (int i = 0; i < n; i++) {
        // Brick i can be safely removed if every brick it supports
        // has at least one other supporter
        bool can_remove = true;
        for (int supported : supports[i]) {
            if (supporters[supported].size() == 1) {
                can_remove = false;
                break;
            }
        }
        if (can_remove) {
            safe_count++;
        }
    }

    return safe_count;
}

int part2(std::vector<Brick>& bricks) {
    auto result = settle_bricks(bricks);
    auto& supports = result.supports;
    auto& supporters = result.supporters;

    int total_falls = 0;
    int n = bricks.size();

    for (int i = 0; i < n; i++) {
        // Simulate removing brick i and count chain reaction
        // BFS to find all bricks that would fall
        std::set<int> falling;
        falling.insert(i);
        std::queue<int> queue;
        queue.push(i);

        while (!queue.empty()) {
            int brick = queue.front();
            queue.pop();

            // Check all bricks that this brick supports
            for (int supported : supports[brick]) {
                if (falling.count(supported)) continue;

                // This brick falls if all its supporters have fallen
                bool all_fallen = true;
                for (int s : supporters[supported]) {
                    if (!falling.count(s)) {
                        all_fallen = false;
                        break;
                    }
                }

                if (all_fallen) {
                    falling.insert(supported);
                    queue.push(supported);
                }
            }
        }

        // Don't count the initial brick we removed
        total_falls += falling.size() - 1;
    }

    return total_falls;
}

int main() {
    std::string input_path = "../input.txt";
    auto bricks = parse_input(input_path);

    std::cout << "Part 1: " << part1(bricks) << std::endl;
    std::cout << "Part 2: " << part2(bricks) << std::endl;

    return 0;
}
